Do not print telemetry info when metrics are off

stevenbuccini opened the pull request:
Currently, the debug message about how to turn off metrics prints at least once, even if the metrics have already been disabled. Now, the metrics will print once only when metrics are being sent. Resolves #5333.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/semgrep/semgrep?pullRequest=11052) <br/>All committers have signed the CLA.

dijkstracula commented:
Hi, thanks for the contribution, however your patch doesn't compile.  Please ensure `make` and `make test` succeeds.

Thanks!

yosefAlsuhaibani commented:
👋 @stevenbuccini ! Thank you for your contribution, note that the patch above only affects invoking semgrep with the `--experimental` flag (i.e the OCaml port); you might also need to modify https://github.com/semgrep/semgrep/blob/320793da396fd1cb3f31c882dd8171e11359d116/cli/src/semgrep/notifications.py#L18-L33 (which is the codepath that regular invocation of semgrep takes i.e as described in #5333)

+ Feel free to ping me for review/questions.

stevenbuccini committed:
Do not print telemetry info when metrics are off

Currently, the debug message about how to turn off metrics prints at least once, even if the metrics have already been disabled. Now, the metrics will print once only when metrics are being sent.

stevenbuccini committed:
Add OCaml tests to verify metric debug text suppression

Added tests to verify that the metrics debug text is suppressed when:
* explicitly set to off
* when metrics are enabled on subsequent runs

I also explicitly turned metrics on in the verbose output because the
default setting is "auto" and its unclear what the behavior of that
setting is in the test environment.

stevenbuccini commented:
I've fixed the logic bug I had in the OCaml change and added the Python change.

On the OCaml side, I've updated the test suite and pushed those up in a separate commit. I am having difficulty with the `basic output with metrics off` test I added because it only scans 2 files instead of 3. The only difference between this test and the existing `basic output` test is the metrics flag setting. I'm actually not sure where the third file is coming from in the first place because I only see two files added in the test itself. Any ideas? 

I'm having trouble getting the Python tests work nicely on my local machine but I will continue to iterate on those.

stevenbuccini commented:
Hey @yosefAlsuhaibani, this change is ready for review. I could use guidance on the OCaml test flakiness. I could also use your help regenerating some of the Python test outputs. I'm running into the package issues described in #11069 so `make regenerate-tests` includes a lot of polluted output. It looks like the main test output that's affected are:
* `cli/tests/default/e2e/snapshots/test_misc/test_terminal_output/results.txt`
* `cli/tests/default/e2e-other/snapshots/test_download_config/test_new_feature_registry_config/output.txt`

Any thoughts?

yosefAlsuhaibani reviewed (approved):
Thank you for your contribution! see my comment below

yosefAlsuhaibani commented:
@stevenbuccini weird: I've cloned and built your branch and it seems like I got:
```
  Scanning 2 files tracked by git with 1 Code rule:
  Scanning 2 files.
```
I wonder if it's something related to your local dev environment, (I would not be surprised if the test utilities that we use for `Test_scan_subcommand.ml` are system dependent ☹️). 

As for the python error: I am still looking into #11069, I did get the warning to go away with the described fix in #11069; however there is a pytest that fails as it depends on an environment value. I am hoping to get fixes for both by the next release (i.e next week). 

Both of the issues above are related to our testing infra, and I do want to unblock this PR: I could push the snapshot fixes to your fork in the meantime?

stevenbuccini commented:
> @stevenbuccini weird: I've cloned and built your branch and it seems like I got:
> 
> ```
>   Scanning 2 files tracked by git with 1 Code rule:
>   Scanning 2 files.
> ```

For which tests? I actually think that's the correct behavior. The [existing test](https://github.com/semgrep/semgrep/blob/dc9770fdd4cd60a9137df8afc77238b06d0b8614/tests/snapshots/semgrep-core/340d1149ba01/stdxxx) expects the following output despite the top of the file explicitly stating there are only two input files: 
```
┌─────────────┐
│ Scan Status │
└─────────────┘
  Scanning 3 files tracked by git with 1 Code rule:
  Scanning 3 files.
```

I'm not sure what the mystery third file is or where it's coming from.

Feel free to push to this branch, although I've fixed my dependency problem and can regenerate the snapshots now as well.

stevenbuccini commented:
Hey @yosefAlsuhaibani, do you mind taking the lead on landing this since you have the most visibility into some other fixes that are pre-requisites to merging this in?

yosefAlsuhaibani commented:
@stevenbuccini sure: apologizes I got pulled into some other work and this fell through the cracks.

yosefAlsuhaibani committed:
update snapshots

yosefAlsuhaibani committed:
fixup! update snapshots
